### PR TITLE
[Core] Add dynamic chunk size calculation

### PR DESCRIPTION
**Pull request by prashantgupta24**
When doing chunked prefill, calculate the token budget for a single chunk. This dynamically scales the chunk size down as the number of sequences that require prefilling increases. This ensures that a single sequence with a very large prompt to prefill doesn't take the entire remaining token budget, allowing other sequences to prefill and decode concurrently.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/core/scheduler.py [hunk 1]
```diff
@@ -1583,6 +1583,43 @@ def _get_num_lookahead_slots(self, is_prefill: bool,
 
         return self.scheduler_config.num_lookahead_slots
 
+    def _get_budget_for_chunk(self, budget: SchedulingBudget) -> int:
+        """When doing chunked prefill, calculate the token budget for a single
+        chunk. This dynamically scales the chunk size down as the number of
+        sequences that require prefilling increases. This ensures that a single
+        sequence with a very large prompt to prefill doesn't take the entire
+        remaining token budget, allowing other sequences to prefill and decode
+        concurrently."""
```
Comment (comaniac) [inline vllm/core/scheduler.py:1592]: The statement "scales the chunk size down" is kind of confusing. IIUC, we don't actually reduce the chunk size for this batch, but reducing the "allocated token budget" for this request, so `_get_token_budget_for_request` may be a better naming.

File: vllm/core/scheduler.py [hunk 2]
```diff
@@ -1583,6 +1583,43 @@ def _get_num_lookahead_slots(self, is_prefill: bool,
...
+
+        # First get the number of sequences that require prefill
+        prefilling_seqs = 0
+        for i in range(len(self.running)):
+            for seq in self.running[i].seqs:
+                if seq.data.stage == SequenceStage.PREFILL:
+                    prefilling_seqs += 1
+        # Queued requests will also need to prefill
+        prefilling_seqs += len(self.waiting)
```
Comment (comaniac) [inline vllm/core/scheduler.py:1601]: Since this function will be called by scheduler in every step, I'm afraid this loop would introduce latency overheads especially with high QPS.

File: vllm/core/scheduler.py [hunk 3]
```diff
@@ -1583,6 +1583,43 @@ def _get_num_lookahead_slots(self, is_prefill: bool,
...
+        # need to prefill
+        chunk_size = int(remaining_token_budget / prefilling_seqs)
+        # Ensure the chunk size is at least the minimum configured by the
+        # user, to limit the number of requests doing prefill
+        chunk_size = max(chunk_size, self.scheduler_config.min_chunk_size)
+        # And cap that at our actual budget so we don't spend tokens we
+        # don't have.
+        chunk_size = min(remaining_token_budget, chunk_size)
+
+        logger.debug("Prefill chunk size: %d", chunk_size)
```
Comment (comaniac) [inline vllm/core/scheduler.py:1620]: Do not use any logging. Even `logger.debug` would introduce latency overheads in the critical path.

File: vllm/core/scheduler.py [hunk 4]
```diff
@@ -1583,6 +1583,43 @@ def _get_num_lookahead_slots(self, is_prefill: bool,
...
+
+        # calculate a chunk size that shares it evenly across sequences that
+        # need to prefill
+        chunk_size = int(remaining_token_budget / prefilling_seqs)
+        # Ensure the chunk size is at least the minimum configured by the
+        # user, to limit the number of requests doing prefill
+        chunk_size = max(chunk_size, self.scheduler_config.min_chunk_size)
+        # And cap that at our actual budget so we don't spend tokens we
+        # don't have.
+        chunk_size = min(remaining_token_budget, chunk_size)
```
Comment (comaniac) [inline vllm/core/scheduler.py:1618]: Again I feel in high QPS the overhead of this logic would be large. Specifically, when there are many prefill requests, you mostly would just allocate min_chunk_size, making this calculation not effective.

**Review by comaniac — COMMENTED** [review]
Thanks for the PR. While I understand that this approach attempts to balance the TTFT of all requests, I have the following concerns:

1. This results in many requests in the partial prefill stage. For example, you may have 10 requests with 1k prompt, and you will schedule all of them to process a chunk in each step. However, we now allocate kv blocks for an entire prompt when scheduling a request, meaning that we will allocate 10k/16 blocks in the first step. Even assuming we have more than 10k/16 blocks available, this may result in lots of preemptions once all 10 requests get to decoding stage.
2. The implementation seems to have some overheads. Please be aware that scheduler is in the critical path of TTFT and ITL, so even 1 ms is a huge overhead.
3. I don't think this approach benefits throughputs, so for the high throughput scenario, people would not use this feature (`min_chunk_size=None`). Then there's only overheads.

In short, I'm worry about the overall performance for the scenarios that cannot benefit from this approach. A solid benchmark could be a good reference to help proceed.

**Comment by njhill** [comment]
I haven't looked at the impl/logic closely but instead of treating all prefills equally perhaps we can admit new ones based on the input size being smaller than some fraction of the remaining prefill tokens of existing prefill reqs in the batch. This would address the core problem while I think avoiding most of the issues @comaniac raised. And in theory could even have some benefit to throughput since more reqs would get to decode stage faster?

**Comment by joerunde** [comment]
> Thanks for the PR. While I understand that this approach attempts to balance the TTFT of all requests, I have the following concerns:
> 
> 1. This results in many requests in the partial prefill stage. For example, you may have 10 requests with 1k prompt, and you will schedule all of them to process a chunk in each step. However, we now allocate kv blocks for an entire prompt when scheduling a request, meaning that we will allocate 10k/16 blocks in the first step. Even assuming we have more than 10k/16 blocks available, this may result in lots of preemptions once all 10 requests get to decoding stage.
> 2. The implementation seems to have some overheads. Please be aware that scheduler is in the critical path of TTFT and ITL, so even 1 ms is a huge overhead.
> 3. I don't think this approach benefits throughputs, so for the high throughput scenario, people would not use this feature (`min_chunk_size=None`). Then there's only overheads.
> 
> In short, I'm worry about the overall performance for the scenarios that cannot benefit from this approach. A solid benchmark could be a good reference to help proceed.

Yeah, all good points!

1: I wouldn't expect `min_chunk_size` to be configured so small that all 10 requests would start to prefill, e.g.that would mean setting `--max-num-batched-tokens=512 --min-chunk-size=50` and with chunks that small, I think prefill would just take a lot longer in general. Setting those to 512/256 to allow two requests to prefill at a time would be more reasonable. I agree we don't want to be reckless with blowing up the KV cache, but we're trying out these changes because a customer is frequently hitting an issue where a single 130k context prompt is blocking the queue during prefill for a solid minute, while only using 10% of the KV cache.

2/3: Yeah we definitely need to not do any extra chunk size calculations if min_chunk_size is unset, and we can clean up / cache some calculations to decrease overhead too.

File: vllm/core/scheduler.py (hunk 2 above)
Comment (joerunde) [reply vllm/core/scheduler.py:1601]: 👍  It should be pretty simple to do something much faster like keep either a set or running counter around: `self.running_prefill_seqs`

**Comment by joerunde** [comment]
Re: benchmarking, we created a dataset with super high variance in the prompt lengths- 75% prompts with tens of tokens and 25% prompts with thousands of tokens. This change is currently faster on TTFT but slower on ITL, so we'll work on reducing that overhead and see if we can beat both.

**Comment by comaniac** [comment]
FYI: @rickyyx found an issue with multiple partial prefills on the fly. Particularly there's an assumption in the sampler that we will only have one partial prefill request in a batch. I'm a bit worry this assumption may be everywhere so we have to be careful.

**Comment by rickyyx** [comment]
> FYI: @rickyyx found an issue with multiple partial prefills on the fly. Particularly there's an assumption in the sampler that we will only have one partial prefill request in a batch. I'm a bit worry this assumption may be everywhere so we have to be careful.

Yeah, specifically for the sampler one I think it's assumed we would only have one partial prefill(that doesn't require prompt logprobes): https://github.com/vllm-project/vllm/blob/d58268c56a8ee0eb01c30e7ab7c07c934e1791c2/vllm/model_executor/layers/sampler.py#L1267-L1270

All inputs to the `zip` were assumed to have equal len, but when there are multiple partial prefills, this assumption might not hold true.

**Comment by njhill** [comment]
Closing because this was superseded by https://github.com/vllm-project/vllm/pull/10235.
